### ISSUE TITLE
Header Issues

### ISSUE THREAD

**Issue by Phil1216** [msys2/MINGW-packages#8796]
I'm developing a project using MSYS2 and MinGW64 that involves OpenGL. It was compiling just fine until recently, and now I'm getting some errors that look like this.

In file included from /mingw64/include/GL/glew.h:1219,
                 from GameModels.h:24,
                 from EntityGetter.h:17,
                 from EntityGetter.cpp:3:
/usr/include/w32api/GL/glu.h:68:78: error: expected ')' before '*' token
   68 | void APIENTRY gluQuadricCallback(GLUquadric *qobj,GLenum which,void (CALLBACK *fn)());
      |                                                                     ~        ^~
      |                                                                              )
/usr/include/w32api/GL/glu.h:68:78: error: expected ')' before '*' token
   68 | void APIENTRY gluQuadricCallback(GLUquadric *qobj,GLenum which,void (CALLBACK *fn)());
      |                                 ~                                            ^~
      |                                                                              )
/usr/include/w32api/GL/glu.h:68:79: error: expected initializer before '*' token
   68 | void APIENTRY gluQuadricCallback(GLUquadric *qobj,GLenum which,void (CALLBACK *fn)());
      |                                                                               ^
(I omitted about a dozen similar errors with more functions inside the glu.h header)

https://www.reddit.com/r/opengl/comments/lhxogx/help_with_gluh/

I found someone else who had the same problem on reddit and found out that "glu.h requires that CALLBACK be defined to either __stdcall or nothing." Defining CALLBACK within my code worked fixed these errors, but then I realized that there are two different glu.h files, one associated with MSYS2 and one associated with MinGW64 (notice the paths above).

I started having these issues after I downloaded/updated some MSYS packages using pacman, so I wonder if something changed in the update or if the MSYS2 glu.h didn't exist previously.

I know there might be a way to use linker flags or compiler flags to fix this problem, or maybe I could try to uninstall the MSYS2 version of the file, but since this appears to be an instance where there's friction between MSYS2 and MinGW64, I thought you should be aware.

**Comment by Phil1216** [comment]
Okay, the minimal code is literally to create an empty main function and then try to import glew.h like this:

#include <cstdlib>
#include "GL/glew.h"

using namespace std;

int main(int argc, char** argv) {
    
    return 0;
}

I configured Netbeans to include the flag -I/mingw64/include, and this is the command that Netbeans generated.

g++ -m64 -I/mingw64/include   -c -g -Wall -s -std=c++11 -MMD -MP -MF "build/Debug/MinGW-Windows/main.o.d" -o build/Debug/MinGW-Windows/main.o main.cpp

I'm also using these options for my linker, but make doesn't even get to that point.

-L/mingw64/lib -lglew32 -lopengl32

**Comment by Phil1216** [comment]
If I define CALLBACK myself then suddenly everything works.

#include <cstdlib>
#define CALLBACK
#include "GL/glew.h"

using namespace std;

int main(int argc, char** argv) {
    
    return 0;
}

**Comment by Phil1216** [comment]
Hey Biswa96, has there been any progress on figuring this out?

**Comment by Biswa96** [comment]
I can not reproduce the issue. BTW, are you mixing msys2 and mingw environment? The output contains both ` /mingw64/include/GL/glew.h` and `/usr/include/w32api/GL/glu.h` path. The later one is from msys2/cygwin environment.

**Comment by Phil1216** [comment]
I'm not intentionally maxing the environments. I'm just including glew.h and my compiling tools seem to be pulling headers from both msys2 and mingw64. As I said I remember installing some additional libraries in msys2 using pacman shortly before I started having these issues.

**Comment by Phil1216** [comment]
Here I am years later working on a new project where I decided to get a fresh install of msys2. I would be remiss if I didn't document what I found.

Turns out the problem was that I was indeed mixing environments. I was using make from msys2 (ie under usr/bin) and g++ from mingw64. According to [this stack overflow issue](https://stackoverflow.com/questions/51755089/where-is-make-on-msys2-mingw-w64) mingw64 packages make as mingw32-make.exe rather than make.exe. That led me to be confused about why mingw64 didn't have make, and so I foolishly used the one in usr/bin.

After fixing my environment and using `mklink make mingw32-make.exe` the issue went away.
